File: 725.分隔链表/splitListToParts.go
package problem0725

type ListNode struct {
	Val  int
	Next *ListNode
}

func splitListToParts(root *ListNode, k int) []*ListNode {
	// 取得链表长度
	length := 0
	for head := root; head != nil; head = head.Next {
		length++
	}
	// part代表每一份最低元素个数，remain代表剩余份数，给前remain个加1
	part, remain := length/k, length%k
	result := make([]*ListNode, 0)
	left, right := root, root
	// 分成k份, 每份part个，前remain个多分一个, 实现boolToNumber简化代码
	for i := 0; i < k; i++ {
		for j := 1; j < part+boolToNumber(i < remain); j++ {
			right = right.Next
		}
		if right == nil {
			result = append(result, nil)
		} else {
			next := right.Next
			right.Next = nil
			result = append(result, left)
			left, right = next, next
		}
	}
	return result
}

func boolToNumber(status bool) int {
	if status {
		return 1
	}
	return 0
}

File: 725.分隔链表/splitListToParts.js
// 分隔链表
// 给定一个头结点为 root 的链表, 编写一个函数以将链表分隔为 k 个连续的部分。
// 每部分的长度应该尽可能的相等: 任意两部分的长度差距不能超过 1，也就是说可能有些部分为 null。前面部分的长度大于等于后面部分的长度。
// root 的长度范围： [0, 1000]
// k 的取值范围： [1, 50]
function splitListToParts(root, k) {
	// 找到链表长度
	let length = 0
	for (let cur = root; cur != null; cur = cur.next) {
		length++
	}
	// 计算k份，每份最少个数，前remain份补1
	const minShare = Math.floor(length / k)
	const remain = length % k
	const result = []
	let subHead = root
	let subTail = root
	// 此时就把份数的数组构建好了
	for (let i = 1; i <= k; i++) {
		const subLength = i <= remain ? minShare+1 : minShare
		// 注意j从1开始，因为subHead=subTail，距离就为1
		for (let j = 1; j < subLength; j++) {
			// 此处不会出现subTail为null的情况才对
			subTail = subTail.next
		}
		if (subTail === null) {
			result.push(subHead)
		} else {
			const next = subTail.next
			subTail.next = null
			result.push(subHead)
			subHead = next
			subTail = next
		}
	}
	return result
}